mitodl/ocw-studio
# Allow editing and publishing of test sites

### What are the relevant tickets?
Closes https://github.com/mitodl/ocw-studio/issues/2112

### Description (What does it do?)
This PR makes some changes that allows editing of test sites directly from the `ocw-studio` sites UI and ensures that any publishing action pushes the output up to the test bucket instead of the draft / live buckets. Instead of 2 specific env variables for the test sites (`OCW_WWW_TEST_SLUG` and `OCW_COURSE_TEST_SLUG`) we now have `TEST_ROOT_WEBSITE_NAME` and `OCW_TEST_SITE_SLUGS`. The `OCW_TEST_SITE_SLUGS` variable is intended to be a list of strings, allowing the specification of any number of test sites to be included in the end to end test pipeline. In `e2e_test_site_pipeline.py`, these sites are iterated and pipeline steps are generated based on them and added to the `across`. The logic determining what to override if a site is a test site was moved directly into `site_pipeline.py`. Whether or not a site is a test site is determined simply by seeing if the current site's `name` is in `settings.OCW_TEST_SITE_SLUGS`. If this is true;

 - `noindex` is set to `True`
 - `web_bucket` is set to `settings.AWS_TEST_BUCKET_NAME`
 - `offline_bucket` is set to `settings.AWS_OFFLINE_TEST_BUCKET_NAME`
 - `static_api_url` is set to `settings.STATIC_API_BASE_URL_TEST` (or the `DEV_TEST_URL` constant, if `is_dev` is true)
 - `resource_base_url` is set to `static_api_url`
 - `sitemap_domain` is set to the parsed domain of `static_api_url`
 - `ocw_studio_url` is set to `static_api_url` if the test website is the root test website, otherwise it's a blank string
 - If the site is the root test website, `url_path` is set to the website's `name` property

### How can this be tested?
 - Make sure you have done the basic setup of `ocw-studio` and are configured for publishing courses locally as well as running the end to end test pipeline locally. Read more here: https://github.com/mitodl/ocw-studio/?tab=readme-ov-file#end-to-end-testing-of-site-pipelines
 - Make sure you have the following new env variables set in your `.env` file:
```
TEST_ROOT_WEBSITE_NAME = "ocw-ci-test-www"
OCW_TEST_SITE_SLUGS=["ocw-ci-test-www","ocw-ci-test-course"]
OCW_STUDIO_TEST_URL=http://localhost:8046/
```
 - Go to http://localhost:8043/sites/ocw-ci-test-www and publish the site live, verifying that the publish URL in the drawer shows http://localhost:8046/. When the pipeline succeeds, browse here and verify that the page loads. Images may not load in your browser, because the HTML is pointing to the internal Docker IP of the nginx server. This is done so that when the end to end test pipeline is running within Concourse, it can load the page properly.
 - Do the same at http://localhost:8043/sites/ocw-ci-test-course, but the URL in the publish drawer should be http://localhost:8046/courses/ocw-ci-test-course.
 - Upsert the end to end test pipeline with `docker compose exec web ./manage.py upsert_e2e_test_pipeline`
 - Browse to http://localhost:8080 and locate the end to end test pipeline. Unpause and run it. The pipeline should succeed.
 - Upsert the mass build pipeline with `docker compose exec web ./manage.py upsert_mass_build_pipeline`
 - Make sure you have the [`fly` CLI](https://concourse-ci.org/fly.html) installed locally, and have used `fly login` to add your locally running Concourse instance as a target
 - Run the following commands and ensure that you get nothing back from either of them:
```
fly -t local get-pipeline -p 'mass-build-sites/offline:false,prefix:"",projects_branch:main,starter:"",themes_branch:main,version:draft' | grep ocw-ci-test-www
fly -t local get-pipeline -p 'mass-build-sites/offline:false,prefix:"",projects_branch:main,starter:"",themes_branch:main,version:draft' | grep ocw-ci-test-course
```
 - Pick any non-test website and go to its editor, for example: http://localhost:8043/sites/6-801-machine-vision-fall-2020/
 - Refresh the pipeline for that site i.e. `docker compose exec web ./manage.py backpopulate_pipelines --filter 6-801-machine-vision-fall-2020`
 - Back on the `ocw-studio` page publish the site live, waiting for it to succeed
 - Click the link in the publish drawer to browse the site, which should be like http://localhost:8045/courses/6-801-machine-vision-fall-2020/
